### DISCUSSION TITLE
Quarto doesnt use env vars to build the URL

### DISCUSSION THREAD

**Discussion by ssb-jnk** [quarto-dev/quarto-cli#9998] [Q&A] [answered]
### Description

Greetings. I have a jupyterlab-user-container running on kubernetes, which has jupyter-server-proxy installed, and that makes it so /proxy/portnr is accessible. So that means i can access my Quarto app on that url and port. Problem is, since i dont run jupyterhub, the /user path is not valid.

If i dont set JUPYTERHUB_USER then it wont build the path.
````
export JUPYTERHUB_SERVICE_PREFIX=""
export JUPYTERHUB_USER=""
export JUPYTER_SERVER_URL=https://user-user-593555-0.lab.app-dev.domain.com/
export JUPYTERHUB_HTTP_REFERER=https://user-user-593555-0.lab.app-dev.domain.com/
````

````
quarto preview .
Preparing to preview

Watching files for changes
Browse at http://localhost:4030/
````

Has anyone got a similar issue as me, or know how to fix? Any help is greatly appreciated.

**Comment by mcanouil** [comment]
You can set environment variable to be used by Quarto in `_environement`. See <https://quarto.org/docs/projects/environment.html#environment-file>.

**Reply by ssb-jnk** [reply]
The URL for the jupyterlab-container is somewhat random at launch (containerID), and so the easiest would be to use the environment variables i mentioned above.

I did make it work, but the environment variables must have values, but since i dont run jupyterhub, /user is not a valid path. The / path is the user already, and thus i should not need to put /user/myusername. I want the path to be /proxy/5851.

This is output, if i leave JUPYTERHUB_USER empty then it will use localhost url
https://user-user-593555-0.lab.app-dev.domain.com/user/myuser/proxy/5851/

I could possibly create a startup script which dynamically on launch populates the _environemnt file you mentioned, but it feels a little overkill for something that should be simple.

**Reply by mcanouil** [reply]
Unfortunately, I am not familiar at all with this workflow thus the only mention of `_environment` file.

**Reply by mcanouil** [reply]
Quarto itself cannot set environment variables but can access them (without computation) using the shortcode: `{{< env YOUR_ENV_VARIABLE >}}`.

You could possibly use that and the preview configuration: https://quarto.org/docs/reference/projects/options.html#preview

**Comment by cderv** [comment]
@ssb-jnk can you clarify what you are expecting from Quarto exactly ? 

You would like the `quarto preview` command to read which env var ?  Or are you saying url is not built as you expect ? 

I would like to understand really the problem encountered and use case so that we can think of right solution (which sometimes can be different that first one thought off). 

Thanks

**Reply by ssb-jnk** [reply]
I apologize if i was unclear. So when i run quarto preview i want the URL generated to point to /proxy/pornr, just like quarto does for Jupyterhub(/user/proxy/portnr), RStudio(/p/portnr) and VSCode(/some/path). Since i run jupyterhub i tried running quarto with the jupyterhub env vars, JUPYTERHUB_HTTP_REFERER, JUPYTERHUB_USER, etc. But problem is quarto points the url to /user/myuser/proxy/portnr, and since i run jupyterlab and not jupyterhub behind the scenes, then /user does not exist. I want to be able to set a environment variable such as 

JUPYERLAB_HTTP_REFERER, and not having to set up JUPYTERHUB_USER (which is required with JUPYTERHUB_HTTP_REFERER  otherwise url will not be generated), since in jupyterlab there is only 1 user it doesnt make sence to have /user endpoint in the generated url.

I am saying the url is not built as expected, yes.

URL generated:
https://user-user1-593555-0.lab-dev.domain.com/user/money/proxy/5851

What i want:
https://user-user1-593555-0.lab-dev.domain.com/proxy/5851

Here it is generated, and it adds /user/someuser to the url generated. https://github.com/quarto-dev/quarto-cli/blob/9b77dcabfefe0d5a9db8dba983da697d797f16f5/src/core/previewurl.ts#L77

**Reply by ssb-jnk** [reply, marked as answer]
Or maybe i can try to configure jupyter to have the path /user/somestaticuser as default path. I think that might solve it.

**Reply by mcanouil** [reply]
It should also be possible to allow the settings to be set under `project.preview` as `server` and `port`.

**Reply by cderv** [reply]
Thanks a lot for clarifying @ssb-jnk ! I know see what you mean. 

We do indeed detect we are in JupyterHub by cheking `JUPYTERHUB_SERVICE_PREFIX` and `JUPYTERHUB_USER` which need to be set and to not empty value. 
https://github.com/quarto-dev/quarto-cli/blob/50b4cf15a0be98a439773ffe923e1e284301879b/src/core/platform.ts#L66-L68

This was tighten up in b2628d318d9abb03fc9c8655cb6a1b8bda0fc2ff for some reason I don't know

Only when those two variable are set to not empty value, quarto will build the URL consulting `JUPYTERHUB_HTTP_REFERER`
https://github.com/quarto-dev/quarto-cli/blob/50b4cf15a0be98a439773ffe923e1e284301879b/src/core/previewurl.ts#L76-L85

This `JUPYTER_USER` env var seems like it should be set in JupyterHub context https://jupyterhub.readthedocs.io/en/stable/reference/spawners.html#environment-variables-and-command-line-arguments
Same for `JUPYTERHUB_SERVICE_PREFIX` that we uses also to detect we are running in JupyterHub. 

How come they are not set in your context ? Is this problematic to set them ? 

Not setting them will also have impact on other command like `quarto publish` or `quarto serve` as they won't detect either it is running on a server

I am trying to understand if the check is too tight and we could safely make it looser. Like checking `JUPYTER_SERVER_URL` or / and `JUPYTERHUB_HTTP_REFERER` instead of USER
